lib/rubocop/cop/lint/unneeded_require_statement.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Checks for unnecessary require statement.
+      #
+      # The following features are unnecessary require statement because
+      # they are already loaded.
+      #
+      # % ruby -e 'p $LOADED_FEATURES.reject { |feature| %r|/| =~ feature }'
+      # ["enumerator.so", "thread.rb", "rational.so", "complex.so"]
+      #
+      # @example
+      #   # bad
+      #   require 'unloaded_feature'
+      #   require 'thread'
+      #
+      #   # good
+      #   require 'unloaded_feature'
+      class UnneededRequireStatement < Cop
+        MSG = 'Remove unnecessary require statement.'.freeze
+
+        LOADED_FEATURES = $LOADED_FEATURES.reject { |feature| %r{/} =~ feature }
```

Comment:
Won't this be `LOADED_FEATURES` of RuboCop? How does inspecting it help us figure out what's loaded in the file we're analyzing?

Reply:
I'm sorry if the answer is different from the intent of the question.

The values returned by `$LOADED_FEATURES` depends on Ruby version.

```console
% rbenv local 2.0.0-p648 && ruby -e 'p $LOADED_FEATURES.reject { |feature| %r|/| =~ feature }'
["enumerator.so"]
% rbenv local 2.1.10 && ruby -e 'p $LOADED_FEATURES.reject { |feature| %r|/| =~ feature }'
["enumerator.so", "thread.rb"]
% rbenv local 2.2.8 && ruby -e 'p $LOADED_FEATURES.reject { |feature| %r|/| =~ feature }'
["enumerator.so", "rational.so", "complex.so", "thread.rb"]
% rbenv local 2.3.5 && ruby -e 'p $LOADED_FEATURES.reject { |feature| %r|/| =~ feature }'
["enumerator.so", "thread.rb", "rational.so", "complex.so"]
% rbenv local 2.4.2 && ruby -e 'p $LOADED_FEATURES.reject { |feature| %r|/| =~ feature }'
["enumerator.so", "thread.rb", "rational.so", "complex.so"]
% rbenv local 2.5.0-dev && ruby -ve 'p $LOADED_FEATURES.reject { |feature| %r|/| =~ feature }'
ruby 2.5.0dev (2017-09-18 trunk 59955) [x86_64-darwin13]
["enumerator.so", "thread.rb", "rational.so", "complex.so"]
```

Due to this difference, depending on the runtime Ruby version of RuboCop, there is a risk that erroneously removing require statement.

For example, running this cop in Ruby 2.4.2 removes `require 'thread`'. However, running that code in Ruby 2.0.0 will result in an error with `require 'thread' missing.

The following is a suggestion to solve the above risk.

For Ruby 2.2 and above, four are returned `["enumerator.so", "rational.so", "complex.so", "thread.rb"]` . These are dependent on the inside of Ruby. It is thought that it will not decrease even if there will be more in the future.

So this PR set `minimum_target_ruby_version` to 2.2 and fix `["enumerator.so", "rational.so", "complex.so", "thread.rb"]` as an array.

Commit 65640f3 is this change.

Reply:
Actually, you should run these commands with `--disable-gems` if you want to know what Ruby loads, and not what RubyGems loads.
For instance, `thread` is not required without RubyGems in 2.2, and it's likely better to not rely on that side effect:
```
$ ruby -v --disable-gems -e 'p $LOADED_FEATURES.reject { |feature| %r|/| =~ feature }'
ruby 2.2.7p470 (2017-03-28 revision 58194) [x86_64-linux]
["enumerator.so", "rational.so", "complex.so"]
$ ruby -e 'p Queue'
Thread::Queue
$ ruby --disable-gems -e 'p Queue' 
-e:1:in `<main>': uninitialized constant Queue (NameError)
```

Reply:
Thanks for the feedback. Certainly it seems better to make some changes on require statement of `thread`. Could you please open an issue for this use case?